chain/vm/fvm.go

```diff
@@ -427,3 +537,110 @@ func (vm *FVM) ApplyImplicitMessage(ctx context.Context, cmsg *types.Message) (*
 func (vm *FVM) Flush(ctx context.Context) (cid.Cid, error) {
 	return vm.fvm.Flush()
 }
+
+type dualExecutionFVM struct {
+	main  *FVM
+	debug *FVM
+}
+
+var _ Interface = (*dualExecutionFVM)(nil)
+
+func NewDualExecutionFVM(ctx context.Context, opts *VMOpts) (Interface, error) {
+	main, err := NewFVM(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	debug, err := NewDebugFVM(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dualExecutionFVM{
+		main:  main,
+		debug: debug,
+	}, nil
+}
+
+func (vm *dualExecutionFVM) ApplyMessage(ctx context.Context, cmsg types.ChainMsg) (ret *ApplyRet, err error) {
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		ret, err = vm.main.ApplyMessage(ctx, cmsg)
+	}()
+
+	go func() {
+		defer wg.Done()
+		if _, err := vm.debug.ApplyMessage(ctx, cmsg); err != nil {
+			log.Errorf("debug execution failed: %w", err)
+		}
+	}()
```

Comment:
So... this doesn't work as the debug messages will have different gas fees. That means:

1. The balances will be different.
2. Some messages may just fail.

This will lead to really annoying and hard to debug issues. We _can_ merge this as a WIP_JUST_TESTING patch, but we need to make that clear.

The right way to do this would be to:

1. Apply in debug mode (disabling gas accounting?). Balances will be correct because we charge for gas by charging for the gas _limit_ up-front, then refunding any leftovers. So the balance will be correct for the duration of the message execution, just not after the message is done executing.
2. Revert.
3. Apply normally.
4. Go back to 1 for the next message.

But:

1. We can't currently perform that revert.
2. That is even more likely to cause problems...

I think the answer here is to leave this as a super experimental feature, but be _very_ clear that this is _just_ for debugging and absolutely not to be relied on.

Reply:
Yeah, I'll throw on a disclaimer.

Honestly, the biggest use I envision getting out of this is actually intentionally failing messages, and using the actor error string to convey debug information (eg. adding `actor_error!("i reached this if branch because the miner power is non-zero");` to the miner_actor).